### PR TITLE
680 hiv testing encounter update

### LINKED ISSUE #680: HIV Testing and Counselling > Encounter form: forms should take the HTC client intake form for the basis
[HTC client intake form](https://docs.google.com/document/d/1-dHq-scMDgHec9gd8k8pDFGoW3RA_J0H/edit)

Most of the fields in Section A - Demographics data are already captured at Patient Registration or Enrolment level. Checking with local team for the following: Partner's HIV Status, Referred from, was client referred by a peer educator ?

Note: numbering is just to state the ordering and dependencies, no need to number each property. 

**For dependencies, would be nice to have a dynamic form (eg. if I have entered "NO" for "Have you been tested for HIV before" then we don't need the dependent property "If yes, what date ? Test Result ?".** 

### Risk Behaviour:

- [x] 1. Have you been tested for HIV before and know your result ? Yes or No (only for tests not captured in mSupply, otherwise we know if patient has already been tested, we can just look at previous encounters)
- [x] 1a If yes, what date ? Test result ? 
- [x] 2. Have you ever experienced any of the following in the past 12 months (refer to the [form](https://docs.google.com/document/d/1-dHq-scMDgHec9gd8k8pDFGoW3RA_J0H/edit))
- [x] 3. How many sexual partners have you had in the last 6 months ? Males and Females
- [x] 4. Risk Group (preferred to Target Population): FSW, MSW, MSM, TG, HR, GP
- [x] 5. GBV Screening outcome: positive, negative, not done, declined
- [x] 5a If GBV screening = positive, referred for GBV services ? Yes or No

### Informed Consent:

- [x] Was pre-test counselling done ? Yes or No
- [x] Was the consent form signed ? Yes or No

note: Could we add the possibility to upload the form ? Will suggest it to Poruan. 

### HIV Testing:

- [x] 1. Date of HIV test done: 
- [x] 2. Confirmed HIV test result: Negative or Positive (note: test results are provided immediately, like RAT for Covid)
- [x] 2a. Did you refer the patient for ART services ?
- [x] 2aa. Where  ? This facility or Another Facility
- [x] 2aaa. if another facility, name of the facility (= list of ART clinics = stores ?)
- [x] 2b. Was the client escorted ? Yes or no
- [x] 3. Post-test counselling done ? Yes or No
- [x] 4. If serodiscordant, was couple counselling provided ? Yes, No, N/A
- [x] 5. Was client screened and referred for other services ? Yes or NO
- [x] 5a. If yes, which one(s): TB, STI, GBV, PEP, Other
- [x] 5aa. If other, specify: 
- [x] 6. Where condoms and/or lubricants offered to the client ? Yes or No
- [x] 6a. If yes number of males condoms provided ? number of female condoms provided ? Lubricants ?
- [x] 8. Did you conduct condom demonstration ? Yes or no

### PR DESCRIPTION
**Pull request by clemens-msupply**
Closes #680 and part of #789 

Includes some bug fixes and shows an error message in the category bubbles:

image

Not pretty but didn't want to spend more time on it right now.

Have the feeling some questions can be grouped further to make it clearer that they belong together...

Also might have to look for more icons.

### CODE REVIEW

**Comment by mark-prins** [comment]
> Includes some bug fixes and shows an error message in the category bubbles:

Initial thought: you could use a Badge on the category bubble to show a number. It wouldn't have the 'missing' text though 🤷

File: client/packages/programs/src/JsonForms/common/components/Array/EnumArray.tsx [hunk 1]
```diff
@@ -76,12 +85,25 @@ export const EnumArrayComponent: FC<EnumArrayControlCustomProps> = ({
...
   schema,
   visible,
+  enabled,
   addItem,
   removeItems,
 }) => {
   const t = useTranslation('common');
   const [removeIndex, setRemoveIndex] = useState<number | undefined>();
 
+  const renderInput = (props: AutocompleteRenderInputParams) => (
```
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/common/components/Array/EnumArray.tsx:95]: can you move this outside of the `EnumArrayComponent` please?
Not really an issue in this case, as the component is not large; it's a good habit to have tho.
Currently the `renderInput` component is recreated on every render, which means that if `EnumArrayComponent` has to re-render, it redraws all child components. If the `renderInput` is created at a higher level, then only the updated elements need to redraw.

File: client/packages/programs/src/JsonForms/common/components/CategorizationTabLayout.tsx [hunk 2]
```diff
@@ -95,6 +99,75 @@ const CategoryLayoutRendererComponent = ({
 };
 const CategoryLayoutRenderer = React.memo(CategoryLayoutRendererComponent);
 
+// Try to extract a more precise error path
+const propertyPathFromError = (error: {
```
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/common/components/CategorizationTabLayout.tsx:103]: Perhaps these two methods would be useful in a `utils` file? They look like useful shared functionality

File: client/packages/programs/src/JsonForms/common/components/CategorizationTabLayout.tsx [hunk 3]
```diff
@@ -95,6 +99,75 @@ const CategoryLayoutRendererComponent = ({
...
+  if (foundPaths.length === 0) {
+    return null;
+  }
+  return (
+    <Typography
+      sx={{
+        color: theme => theme.palette.error.main,
+        backgroundColor: theme => theme.palette.background.login,
+        borderRadius: '16px',
+        padding: '0px 8px',
```
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/common/components/CategorizationTabLayout.tsx:163]: In order to remove hard coded pixel values 🤷 
```suggestion
        borderRadius: 4,
        paddingX: 2,
```

**Review by mark-prins — APPROVED** [review]
👍  LGTM!

File: client/packages/programs/src/JsonForms/common/components/CategorizationTabLayout.tsx (hunk 2 above)
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/common/components/CategorizationTabLayout.tsx:103]: I leave it for another time when I really needed it somewhere else (it was a bit hacky and took longer than I wanted so would like to give it a second thought then...)

**Comment by clemens-msupply** [comment]
> > Includes some bug fixes and shows an error message in the category bubbles:
> 
> Initial thought: you could use a Badge on the category bubble to show a number. It wouldn't have the 'missing' text though shrug

Yes good idea, also though to stick the Badge to the bottom right corner of the bubble...

File: client/packages/programs/src/JsonForms/common/components/Array/EnumArray.tsx (hunk 1 above)
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/common/components/Array/EnumArray.tsx]: done thanks!

**Pull request merged**